src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs

```diff
@@ -200,6 +202,59 @@ internal static IEnumerable<ReplicationBatchItem> GetReplicationItems(DocumentDa
             }
         }
 
+        private class ReplicationBatchState : PulsedEnumerationState<ReplicationBatchItem>
+        {
+            private ReplicationDocumentSender _parent;
+            private ReplicationState _replicationState;
+            public bool WasInterrupted { get; private set; }
+            public long Next { get; private set; }
+
+            public ReplicationBatchState(DocumentsOperationContext context, Size pulseLimit, ReplicationDocumentSender parent, ReplicationState replicationState, long next) : base(context, pulseLimit)
+            {
+                _parent = parent;
+                _replicationState = replicationState;
+                WasInterrupted = false;
+                Next = next;
+            }
+
+            public override void OnMoveNext(ReplicationBatchItem current)
+            {
+                _parent._parent.ForTestingPurposes?.OnDocumentSenderFetchNewItem?.Invoke();
+                _parent._parent.CancellationToken.ThrowIfCancellationRequested();
+                _parent.AssertNoLegacyReplicationViolation(current);
+
+                if (_replicationState.LastTransactionMarker != current.TransactionMarker)
+                {
+                    _replicationState.Item = current;
+                    long next = Next;
+                    if (_parent.CanContinueBatch(_replicationState, ref next) == false)
+                    {
+                        WasInterrupted = true;
+                        return;
+                    }
+
+                    Next = next;
+                    _replicationState.LastTransactionMarker = current.TransactionMarker;
+                }
+            }
+
+            public override bool ShouldPulseTransaction()
+            {
+                if (_replicationState.NumberOfItemsSent == 0)
```

Comment:
why is this only in this scenario? we may want to split the batch in this case anyway, no?
Or is it to ensure transaction atomicity in the replication?

Reply:
We pulse transactions (eliminate and create new transactions) only for the documents that are scanned and not sent in batches when we fetch them.
We cannot split the documents that are supposed to be sent because if we destroy the transaction they are sitting on during the fetch (before creating a new transaction), their blittable will be deleted along with the transaction and we wouldn't be able to send them.

Reply:
When we do the pulse we invalidate all previous data.
For replication we first collect the entire batch and only then send it (because we need to send the number of items we collected. etag, etc.. ).
Other features using the pulsed tx, are write it directly to the stream.

Reply:
Okay, good point on invalidating old references.